Navigation on Browse Mode: Move the action to the leaf menu

scruffian opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
This moves the navigate action from being the onSelect action of the block to being hidden under the ellipsis

## Why?
This action is only relevant on items that link to a page. It also takes users to the Pages section of browse mode, so it makes more sense as a secondary action.

## How?
We no longer need to modify the behaviour of the onSelect action in the List View, so the code becomes simpler - we just move the code to the leaf-more-menu file for this component.

## Testing Instructions
1. Add a navigation block to a template
2. Add a link to a page on your site
3. Add a page link block
4. Add a custom link block
5. Add a post link block

Clicking on the ellipsis menu by the page and post link blocks should open the more menu with the option to "Navigate to [Page X]". The custom link block should not provide the option to navigate.

Also try adding blocks without urls/page set and confirm that these blocks don't show the option to "Navigate to X".

Also confirm that the action to select a page and then navigate to it from the List View does not work.

<img width="597" alt="Screenshot 2023-05-22 at 12 57 57" src="https://github.com/WordPress/gutenberg/assets/275961/c2b8b8bc-c2e9-47e7-ada9-813846ba0a5c">

getdave reviewed (commented):
The change itself seems good. Thank you for working on it 🙇 

The only bug I found was when I added a Category link using the `Category` link variation. It shows the `Navigation to...` item but when clicked it does nothing.

Conceptually however I think users will find this super confusing. We've been teaching them that in list view "click" means "select this item". Now we're adding an edge case where "click" does nothing at all and you have to know about some obscure option to trigger that action.

Whilst I appreciate the sentiment that it might be strange to have click navigation to a page, I feel what we have in this PR is not a good experience and we should reconsider.

jameskoster commented:
The label is a little tricky – "Navigate to X" could be interpreted as viewing the page on the frontend. 

Perhaps something like "View page details" or even "Edit page details"?

Related, but not something to address here... is there a chance "Remove X" could be mistaken for deleting the page? Maybe a more generic "Remove link" would work better?

jasmussen commented:
How about "Go to X"?

jameskoster commented:
I don't know that we need to include the title? It's already clearly visible in the menu item, and we'd need to truncate long titles. Given we removed the block name in the block toolbar's "Delete" action, it may make sense to be consistent here?

richtabor commented:
> Maybe a more generic "Remove link" would work better?

Or "Delete" (re consistency with the block options). 

> How about "Go to X"?

What about "Go to page" or even the classic and familiar "Edit page"? 

I don't think we need to be prescriptive about the editor view you're in (if page details is open, or you're right in the editor). Editing the page starts with the Page Details view.

jameskoster commented:
I thought "Edit page" initially. What put me off is that you can't actually edit anything in the details view yet, and there is a separate "Edit" button there too which makes things a little awkward. 

There _may_ be an argument for separate actions in this menu: 

1. View page details – takes you to the details panel
2. Edit page – takes you directly to the full screen editor

That's probably a bigger discussion though. For now I'd lean towards "Go to page" which simply opens the details panel.

scruffian committed:
Move the go to action to the leaf more menu

scruffian commented:
I rebased this, so it's ready for another review.

jasmussen commented:
Generally looks like this PR is landable, as the core behavior feels like an upgrade. No chevron to indicate drilldown, and the ellipsis menu shows a "go to":

<img width="574" alt="Screenshot 2023-06-19 at 09 49 11" src="https://github.com/WordPress/gutenberg/assets/1204802/0c1891be-00ed-4318-a40a-d4d36e7d0f08">

However there are some glitches in the site view navigation overall, one of which may be worth fixing in this branch. Notably here:

<img width="354" alt="Screenshot 2023-06-19 at 09 48 31" src="https://github.com/WordPress/gutenberg/assets/1204802/7b8790b1-5042-46a8-bd96-852090b9ca5d">

The width of the active bar does not feel balanced left and right, and the ellipsis doesn't line up with the heading ellipsis. I understand that the icons on the left are a bit trickier since there has to be room for the chevron for submenus, but I think with a little CSS that can be fiddled as well, so that those left icons are positioned according to the same type of icons in other  menus. Compare with icons in the top level menu:

<img width="358" alt="Screenshot 2023-06-19 at 09 46 50" src="https://github.com/WordPress/gutenberg/assets/1204802/b9930ff3-235e-4465-84e4-dc02130c6274">

<img width="359" alt="Screenshot 2023-06-19 at 09 46 36" src="https://github.com/WordPress/gutenberg/assets/1204802/68e20796-10ae-4815-93c2-61f5d5e659d5">

---

Almost certainly separate:

<img width="500" alt="Screenshot 2023-06-19 at 09 49 22" src="https://github.com/WordPress/gutenberg/assets/1204802/ae0a9a50-e9a7-4235-a19f-dd62bfa705ce">

Let's not use the red color for the button here. In general I think we should move away from "destructive colors" as they do not work well for color blindness and don't always culturally have the correct significance. In this case the confirm dialog itself is enough validation that you are sure. You can also change the "Confirm" label to "Delete".

There's also a bit of an issue with how multi menu management works. The heuristic is this:

* If there's only a single menu saved, the structure is Design > Navigation [showing menu items directly]
* If there's more than one menu, the structure is Design > Navigation > Name of menu [showing menu items here]

Note the following GIF how this only partially works:

![nav2](https://github.com/WordPress/gutenberg/assets/1204802/a10e6205-d7af-44e5-b6e7-edcea529740d)

In this GIF it shows how clicking "Navigation" skips the "Navigation" section and jumps to the "Header navigation" section, so when you click the "Back" button you go to a "Navigation" section you didn't navigate to, showing a single drilldown for the menu that's saved, like so:

<img width="359" alt="Screenshot 2023-06-19 at 09 58 57" src="https://github.com/WordPress/gutenberg/assets/1204802/0421ce10-b35e-4f2b-9ae0-1fb0f49dcccd">

You should never see this. When there's only one menu, show it directly on this top level of the Navigation section, as per #50396:

<img src="https://user-images.githubusercontent.com/1204802/238354901-4a822f8a-7e7e-46fb-af71-e29afb8052c8.png">

jasmussen commented:
Looks like Jay noted some of the alignment issues already in https://github.com/WordPress/gutenberg/pull/51585#issuecomment-1594891680.

getdave commented:
> You should never see this. When there's only one menu, show it directly on this top level of the Navigation section, as per #50396:

This has been noted in https://github.com/WordPress/gutenberg/issues/51561 and we are fixing in https://github.com/WordPress/gutenberg/pull/51565

> Let's not use the red color for the button here. In general I think we should move away from "destructive colors" as they do not work well for color blindness

If that's the case then let's deprecate the props of the relevant `@wordpress/components` so that we _codify_ the convention 🙏

jasmussen commented:
CC: @ciampo in case you have advice on the color bit.

jasmussen commented:
Oh sorry for the double ping, I just now realize you already created a separate issue! You're popular, Marco!

scruffian commented:
I added https://github.com/WordPress/gutenberg/issues/51690 to track the modal, and added the feedack about the alignment to https://github.com/WordPress/gutenberg/issues/51581.

I'd like to get this one merged and do the other changes in followups.

jasmussen reviewed (approved):
Would appreciate a code check, and thanks for looking at the followups.

draganescu commented:
This PR makes a bug that exists in trunk more visible:

<img width="1315" alt="Screenshot 2023-06-20 at 14 36 31" src="https://github.com/WordPress/gutenberg/assets/107534/d1c9a2a5-7dc4-494a-8cb3-f738d659ad4b">


- [ ] when selecting an item from the navigation a horizontal scrollbar appears

This can also be seen in `trunk` if a custom link is added to a navigation and then selected in site view.

getdave commented:
🥳
